src/utils/Simplify.h

```diff
@@ -0,0 +1,465 @@
+        const coord_t length2_after = vSize2(vertex_position - after_position);
+
+        if(length2_before <= max_resolution * max_resolution && length2_after <= max_resolution * max_resolution) //Both adjacent line segments are short.
+        {
+            //Removing this vertex does little harm. No long lines will be shifted.
+            to_delete[vertex] = true;
+            return;
+        }
+
+        //Otherwise, one edge next to this vertex is longer than max_resolution. The other is shorter.
+        //In this case we want to remove the short edge by replacing it with a vertex where the two surrounding edges intersect.
+        //Find the two line segments surrounding the short edge here ("before" and "after" edges).
+        Point before_from, before_to, after_from, after_to;
+        if(length2_before <= length2_after) //Before is the shorter line.
+        {
+            if(!is_closed && before == 0) //No edge before the short edge.
+            {
+                return; //Edge cannot be deleted without shifting a long edge. Don't remove anything.
+            }
+            const size_t before_before = previousNotDeleted(before, to_delete);
+            before_from = getPosition(polygon[before_before]);
+            before_to = getPosition(polygon[before]);
+            after_from = getPosition(polygon[vertex]);
+            after_to = getPosition(polygon[after]);
+        }
+        else
+        {
+            if(!is_closed && after == polygon.size() - 1) //No edge after the short edge.
+            {
+                return; //Edge cannot be deleted without shifting a long edge. Don't remove anything.
+            }
+            const size_t after_after = nextNotDeleted(after, to_delete);
+            before_from = getPosition(polygon[before]);
+            before_to = getPosition(polygon[vertex]);
+            after_from = getPosition(polygon[after]);
+            after_to = getPosition(polygon[after_after]);
+        }
+        Point intersection;
+        const bool did_intersect = LinearAlg2D::lineLineIntersection(before_from, before_to, after_from, after_to, intersection);
+        if(!did_intersect) //Lines are parallel.
+        {
+            return; //Cannot remove edge without shifting a long edge. Don't remove anything.
+        }
+        const coord_t intersection_deviation = LinearAlg2D::getDist2FromLine(intersection, before_to, after_from);
+        if(intersection_deviation <= max_deviation * max_deviation) //Intersection point doesn't deviate too much. Use it!
+        {
+            to_delete[vertex] = true;
+            if(length2_before <= length2_after)
+            {
+                polygon[before] = createIntersection(polygon[before], intersection, polygon[after]);
+            }
+            else
+            {
+                polygon[after] = createIntersection(polygon[before], intersection, polygon[after]);
+            }
+        }
```

Comment:
Can this (if block above here -- specifically, the one that replaces a vertex with an intersection) create a situation where the displaced vertex (before or after, depending) is already out of the importance queue (`by_importance`) in the calling method, but now, due to being placed at the intersection, in a position where a next simplify with the same parameters would remove this (or alter something else)?

Reply:
For the fixed-width case I can prove that this can't happen in the current implementation.

Crucially, it will only move the vertex such that two conditions hold:
* It will never decrease the deviation of that vertex. The deviation can only increase or remain the same.
* It will never reduce the length of the edges incident to the vertex. The length can only increase.

If a vertex is processed through the importance queue, it means that either it got deleted (and won't be seen again by `nextNotDeleted` etc.) or it had to stay because of its deviation or the length of its incident edges. Since the deviation can only increase, if it had to stay for the previous check it'll still have to stay. Since the edge length can only increase, if it had to stay due to not being below Max Resolution it'll still have to stay.
There's a third case where the vertex initially has two short edges incident, but due to moving the vertex it gets a long edge that is not allowed to shift (so it would need to calculate the intersection again and move the vertex). However this case can't happen for vertices that have already been processed because then it would have to have started with two short edges and so the vertex is already removed unless it has too much deviation.

For the variable-width case I'm less sure of it, because that adds the additional constraint of the max area deviation. The moved vertex gets the width that was halfway the line segment that got removed. But because the two vertices of the edge are also removed, there is no such constraint on the area deviation of the moved vertex. It might increase (w.r.t. either of the edge's vertices) or it might decrease.
I guess it could happen that a vertex replacing an edge gets a lower area deviation, if the vertices adjacent to the edge's incident vertices have a width that happens to be around the average of the width of the removed edge. If they are closer to the average than the edge's vertices themselves are. However I think that this case is rare enough and harmless enough that we can accept it as a shortcoming.